contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
 # ControllerRef proposal
 
-Author: gmarek@
-Last edit: 2016-05-11
-Status: raw
+* Authors: gmarek, enisoc
+* Last edit: 2017-01-25
+* Status: partially implemented
 
 Approvers:
-- [ ] briangrant
-- [ ] dbsmith
+* [ ] briangrant
+* [ ] dbsmith
 
 **Table of Contents**
 
-- [Goal of ControllerReference](#goal-of-setreference)
-- [Non goals](#non-goals)
-- [API and semantic changes](#api-and-semantic-changes)
-- [Upgrade/downgrade procedure](#upgradedowngrade-procedure)
-- [Orphaning/adoption](#orphaningadoption)
-- [Implementation plan (sketch)](#implementation-plan-sketch)
-- [Considered alternatives](#considered-alternatives)
-
-# Goal of ControllerReference
-
-Main goal of `ControllerReference` effort is to solve a problem of overlapping controllers that fight over some resources (e.g. `ReplicaSets` fighting with `ReplicationControllers` over `Pods`), which cause serious [problems](https://github.com/kubernetes/kubernetes/issues/24433) such as exploding memory of Controller Manager.
-
-We don't want to have (just) an in-memory solution, as we don’t want a Controller Manager crash to cause massive changes in object ownership in the system. I.e. we need to persist the information about "owning controller".
-
-Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
-
-# Non goals
-
-Cascading deletion is not a goal of this effort. Cascading deletion will use `ownerReferences`, which is a [separate effort](garbage-collection.md).
-
-`ControllerRef` will extend `OwnerReference` and reuse machinery written for it (GarbageCollector, adoption/orphaning logic).
-
-# API and semantic changes
-
-There will be a new API field in the `OwnerReference` in which we will store an information if given owner is a managing controller:
-
-```
-OwnerReference {
-    …
-    Controller bool
+* [Goals](#goals)
+* [Non-goals](#non-goals)
+* [API](#api)
+* [Behavior](#behavior)
+* [Upgrading](#upgrading)
+* [Implementation](#implementation)
+* [Alternatives](#alternatives)
+
+# Goals
+
+* The main goal of ControllerRef (controller reference) is to solve the problem
+  of controllers that fight over controlled objects due to overlapping selectors
+  (e.g. a ReplicaSet fighting with a ReplicationController over Pods because
+  both controllers have label selectors that match those Pods).
+  Fighting controllers can [destabilize the apiserver](https://github.com/kubernetes/kubernetes/issues/24433),
+  [thrash objects back-and-forth](https://github.com/kubernetes/kubernetes/issues/24152),
+  or [cause controller operations to hang](https://github.com/kubernetes/kubernetes/issues/8598).
```

Comment:
By "controller operations" I just meant an operation on a controller, such as `kubectl delete <controller>`. My reading of the issue is that operation was hanging due to fighting controllers.